spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/ProjectMetricsAutoConfiguration.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2012-2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.boot.actuate.autoconfigure.metrics;
+
+import java.util.function.Supplier;
+
+import io.micrometer.core.instrument.Gauge;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.binder.MeterBinder;
+
+import org.springframework.boot.autoconfigure.AutoConfiguration;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.info.ProjectInfoAutoConfiguration;
+import org.springframework.boot.info.BuildProperties;
+import org.springframework.boot.info.GitProperties;
+import org.springframework.context.annotation.Bean;
+
+/**
+ * Auto-configuration that exposes Git and build information as metrics.
+ *
+ * @author Matthias Friedrich
+ * @since 3.0.0
+ */
+@AutoConfiguration(after = { MetricsAutoConfiguration.class, CompositeMeterRegistryAutoConfiguration.class,
+		ProjectInfoAutoConfiguration.class })
+@ConditionalOnClass(MeterRegistry.class)
+@ConditionalOnBean(MeterRegistry.class)
+public class ProjectMetricsAutoConfiguration {
+
+	@Bean
+	@ConditionalOnBean(GitProperties.class)
+	public MeterBinder gitInfoMetrics(GitProperties props) {
+		return (registry) -> {
+			Gauge.Builder<Supplier<Number>> builder = Gauge.builder("git.info", () -> 1L)
+					.description("Project Git information").strongReference(true);
+
+			addTag(builder, "branch", props.getBranch());
+			addTag(builder, "id", props.getShortCommitId());
+			addTag(builder, "time", props.getCommitTime());
+
+			builder.register(registry);
+		};
+	}
+
+	@Bean
+	@ConditionalOnBean(BuildProperties.class)
+	public MeterBinder buildInfoMetrics(BuildProperties props) {
+		return (registry) -> {
+			Gauge.Builder<Supplier<Number>> builder = Gauge.builder("build.info", () -> 1L)
+					.description("Project build information").strongReference(true);
+
+			addTag(builder, "name", props.getName());
+			addTag(builder, "artifact", props.getArtifact());
+			addTag(builder, "group", props.getGroup());
+			addTag(builder, "version", props.getVersion());
+			addTag(builder, "time", props.getTime());
+
+			builder.register(registry);
+		};
+	}
+
+	private static void addTag(Gauge.Builder<Supplier<Number>> builder, String name, Object value) {
+		if (value != null) {
```

Comment:
This can be a little dangerous. Let's say one app has the all the git info, the other does not. If that happens you will end-up with the same metric name but different set of tags. This can lead to all sorts of issues and some backends does not or does not like to support this (e.g.: [prometheus](https://github.com/prometheus/client_java/issues/696#issuecomment-920353692)). I would return a default value like `unknown` or `N/A` if the value is missing but I would always want to have the same set of keys.

Here's is an example (ServiceA has all the details, ServiceB only has the commit id)
```
git_info{service="serviceA",branch="main",id="cafecafe",time="2022-04-26T18:39:18"} 1.0
git_info{service="serviceB",id="deadbeef"} 1.0
```
This can be problematic (see above), my proposal instead is:
```
git_info{service="serviceA",branch="main",id="cafecafe",time="2022-04-26T18:39:18"} 1.0
git_info{service="serviceB",branch="unknown",id="deadbeef",time="unknown"} 1.0
```

I have a sad but real-life example for this. If you use heroku in a way that it will build your app when you deploy, heroku removes the .git folder so the git plugin will fail. You can [generate](https://github.com/jonatan-ivanov/resourceater/blob/bd0b061c8d70fd3b8330ed4430491cc3a439f2c4/build.gradle#L69) `git.properties` on your own but heroku will only tell you the commit id so the git info will be [incomplete](https://resourceater.herokuapp.com/actuator/info) (this page might take some time to load).

Reply:
If you do this, you can also simplify things above.

```java
Gauge.builder("git.info", () -> 1L)
        .description("Project Git information").strongReference(true);
        .tag("branch", getOrDefault(props.getBranch()));
		.tag("id", getOrDefault(props.getShortCommitId()));
		.tag("time", getOrDefault(props.getCommitTime()));
		.register(registry);
```

Reply:
Sounds good to me!